File: LeetCode/src/dp/OnesAndZeroes.java
package dp;

public class OnesAndZeroes {
    public int findMaxForm(String[] strs, int m, int n) {
        int[][] dp = new int[m + 1][n + 1];
        dp[0][0] = 0;

        for (String str: strs ) {
            int oneCount = 0;
            int zeroCount = 0;

            for (char s : str.toCharArray()) {
                if (s == '0'){
                    zeroCount++;
                } else {
                    oneCount++;
                }
            }

            for (int i = m; i >= zeroCount; i--) {
                for (int j = n; j >= oneCount ; j--) {
                    dp[i][j] = Math.max(dp[i][j], dp[i - zeroCount][j - oneCount] + 1);
                }
            }
        }

        return dp[m][n];
    }

    private int countZeroes(String str) {
        int count = 0;
        for (char c : str.toCharArray()) {
            if (c == '0') {
                count++;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        OnesAndZeroes solution = new OnesAndZeroes();

        // Test case
        String[] strs = {"10", "0001", "111001", "1", "0"};
        int m = 5;
        int n = 3;
        System.out.println(solution.findMaxForm(strs, m, n)); // Output: 4
    }
}

File: LeetCode_Python/dp/knapsack/OnesAndZeroes.py
class OnesAndZeroes:
    def findMaxForm(self, strs: list[str], m: int, n: int) -> int:
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for s in strs:
            zeros = s.count("0")
            ones = s.count("1")

            for i in range(m, zeros - 1, -1):
                for j in range(n, ones - 1, -1):
                    dp[i][j] = max(dp[i][j], dp[i - zeros][j - ones] + 1)

        return dp[m][n]


# Time Complexity: O(L * m * n), where L is the number of strings, m is the maximum number of 0s, and n is the maximum number of 1s
# Space Complexity: O(m * n), for the dp array

if __name__ == "__main__":
    solution = OnesAndZeroes()

    print("Expected Output: 4")
    print(
        "Actual Output:", solution.findMaxForm(["10", "0001", "111001", "1", "0"], 5, 3)
    )

    print("Expected Output: 2")
    print("Actual Output:", solution.findMaxForm(["10", "0", "1"], 1, 1))
